### PR TITLE
First Commit

### PR DESCRIPTION
**Pull request by cs-sujeet-ahirrao**
First Commit

### CODE REVIEW

File: taxii2-feed/operations.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+def max_age(params, ioc):
+    if params.get("expiry") is not None and params.get("expiry") != '':
+        return get_epoch(ioc["valid_from"]) + (params.get("expiry") * 86400)
+    elif "valid_until" in ioc.keys():
+        return get_epoch(ioc["valid_until"])
+    else:
+        return None
+
+
+def tlp(TLP_AMBER, TLP_RED, TLP_WHITE, TLP_GREEN, params):
```
Comment (cs-pooja-singh) [inline taxii2-feed/operations.py]: why upper case for function params?

File: taxii2-feed/operations.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+def stix_spec(ioc, _version, params):
+    return {
+        "type": "indicator",
+        "spec_version": _version,
+        "created": get_epoch(ioc["created"]),
+        "modified": get_epoch(ioc["modified"]),
+        "recordTags": ioc["indicator_types"] if "indicator_types" in ioc else ioc['labels'],
+        "name": ioc["name"],
+        "description": ioc["description"] if "description" in ioc else None,
+        "indicators": extract_artifacts(data=ioc["pattern"]),
```
Comment (cs-pooja-singh) [inline taxii2-feed/operations.py]: is this expected to create an m2m with these indicators?

File: taxii2-feed/operations.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        "valid_from": get_epoch(ioc["valid_from"]),
+        "confidence": params.get("confidence") if params.get("confidence") is not None and params.get(
```
Comment (cs-pooja-singh) [inline taxii2-feed/operations.py]: leave it as blank instead of setting 0 if not specified from source

File: taxii2-feed/operations.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+            "confidence") != '' else 0,
+        "reputation": REPUTATION_MAP.get(params.get("reputation")) if params.get(
```
Comment (cs-pooja-singh) [inline taxii2-feed/operations.py]: leave it as blank instead of setting Suspicious if not specified from source

File: taxii2-feed/constants.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+REPUTATION_MAP = {'Suspicious': '/api/3/picklists/50bfd06c-9aff-4f7d-b6d9-821339e31fe7',
+                  'Malicious': '/api/3/picklists/7074e547-7785-4979-be32-c6d0c863e4bd',
+                  'No Reputation Available': '/api/3/picklists/9a611980-1b5e-4ae9-8062-eb2c0c433cff',
+                  'TBD': '/api/3/picklists/ae98ebc6-beef-4882-9980-1d88fc6d87cd',
+                  'Good': '/api/3/picklists/b19b42aa-aee4-47df-9cda-894537dacb2a'}
+
+TLP_MAP = {'Red': '/api/3/picklists/0472d368-bd15-4f52-a119-d403470cbe43',
+           'Amber': '/api/3/picklists/7bff95b7-6438-4b01-b23a-0fe8cb5b33d3',
+           'Green': '/api/3/picklists/47004ad3-721e-43e0-b729-2ad8ee6441c0',
+           'White': ''}
```
Comment (cs-pooja-singh) [inline taxii2-feed/constants.py]: should map to None. setting and empty value might fail during insertion

File: taxii2-feed/operations.py (hunk 1 above)
Comment (cs-pooja-singh) [reply taxii2-feed/operations.py]: also, is this function used? don't see its usage anywhere

File: taxii2-feed/operations.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        for bundle in as_pages(collection.get_objects, added_after=params.get('added_after') if (
+                params.get('added_after') is not None and params.get(
+            'added_after') != '') else '1970-01-01T00:00:00.000Z', start=params.get('offset'),
+                               per_request=1000):
+            if 'objects' not in bundle.keys():
+                response.append(bundle)
+                break
+            for ioc in bundle["objects"]:
+                if ioc["type"] == "indicator" and "spec_version" in ioc.keys():
+                    response.append(stix_spec(ioc, "2.1", params))
```
Comment (cs-pooja-singh) [inline taxii2-feed/operations.py]: why do we need spec version in response

File: taxii2-feed/connector.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from connectors.core.connector import Connector, get_logger, ConnectorError
+
+from .operations import operations, _check_health
+
+logger = get_logger('taxii2_feed')
```
Comment (cs-pooja-singh) [inline taxii2-feed/connector.py]: can we rename the connector to taxii2-threat-intel-feed

File: taxii2-feed/info.json [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+        "editable": true,
+        "type": "select",
+        "options": [
+          "Basic Authentication",
+          "SSL Two-Way Certificate"
+        ],
+        "onchange": {
+          "Basic Authentication": [
+            {
+              "title": "User Name / API Key",
```
Comment (cs-pooja-singh) [inline taxii2-feed/info.json]: "User Name/API Key"

File: taxii2-feed/info.json [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+              "type": "textarea",
+              "visible": true,
+              "editable": true,
+              "required": true,
+              "tooltip": "Add a key certificate file as text to connect to the TAXII Client.",
+              "description": "Add a key certificate file as text to connect to the TAXII Client."
+            }
+          ]
+        },
+        "tooltip": "Time stamp of when the indicator was last seen in STAXX. Date can be specified in Date format or Relative Time Unit format.",
```
Comment (cs-pooja-singh) [inline taxii2-feed/info.json]: tooltip and description is for some other key?

File: taxii2-feed/info.json [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+          "name": "added_after",
+          "visible": true,
+          "required": false,
+          "editable": true,
+          "type": "datetime",
+          "tooltip": "Filter the result set to only include the items added after the specified time.",
+          "description": "Filter the result set to only include the items added after the specified time."
+        },
+        {
+          "title": "Save Response in File",
```
Comment (cs-pooja-singh) [inline taxii2-feed/info.json]: to  file

File: taxii2-feed/operations.py (hunk 1 above)
Comment (cs-sujeet-ahirrao) [reply taxii2-feed/operations.py]: Yes, mam, That is not more used ... deleted

File: taxii2-feed/operations.py (hunk 6 above)
Comment (cs-sujeet-ahirrao) [reply taxii2-feed/operations.py]: format of 2.0 and 2.1 is different that's why I added that one

**Review by cs-pooja-singh — APPROVED** [review]

**Pull request merged**